### PR TITLE
Fix #1168: add ability to specify request pass through behavior for …

### LINKED ISSUE #1168: APIG passthroughBehavior property
There is a new functionality in APIG that allows you to define how an APIG method will behave if the incoming content-type does not match any defined `requestTemplate` content-type.
The behavior can be defined in the `passthroughBehavior` REST API property when creating or updating an APIG method resource.
Allowed values are: `WHEN_NO_MATCH`, `WHEN_NO_TEMPLATES` and `NEVER`.
Serverless should support setting this property as the default behavior is to pass the raw body to the attached lambda which is not perfect especially for API methods that should only accept one specific content type (e.g. `application/x-www-urlencoded`).
See here https://docs.aws.amazon.com/apigateway/api-reference/resource/integration/

### PR DESCRIPTION
**Pull request by anreved**
***Implementing Issue:*** #1168 

## What did you implement:
Added ability to specify a pass-through behavior for API Gateway requests.  It requires that you select one of the valid methods (NEVER, WHEN_NO_MATCH, WHEN_NO_TEMPLATES), and uses NEVER as the default value if none was specified.

## How did you implement it:
I updated the apiGateway events method processing to include a check for the "passThrough" configuration.  If none specified, it will use the default value.  If an invalid value is passed, an error will be thrown.

## How can we verify it:
Update your function config under events.http.request to include a "passThrough" value.

```yml
functions:
  create:
    handler: posts.create
    events:
      - http:
          method: get
          path: whatever
          request:
            passThrough: WHEN_NO_TEMPLATES
```

You can verify that the resulting CloudFormation template under Integration -> PassthroughBehavior has the correct value.
 
Once deployed, you can confirm in the API Gateway dashboard under that method -> Integration Request -> Body Mapping Templates.

## Todos:

- [x] Write tests
- [x] Write docs
- [x] Fix linting errors
- [x] Provide verification config/commands/resources

…API Gateway

### CODE REVIEW

File: lib/plugins/aws/deploy/compile/events/apiGateway/README.md [hunk 1]
```diff
@@ -96,6 +96,25 @@ functions:
 **Note:** The templates are defined as plain text here. However you can also reference an external file with the help of
 the `${file(templatefile)}` syntax.
 
+#### Pass Through Behavior
+API Gateway provides multiple ways to handle requests where the Content-Type header does not match any of the specified mapping templates.  If not specified, a value of *NEVER* will be used.
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/compile/events/apiGateway/README.md]: Can you also mention the `415 - Unsupported Media Type` message users will get when there is no template defined for their content-type. This way if they search the docs its easy for them to find it this setting and set it.

**Comment by flomotlik** [comment]
Hey @anreved , thanks for the PR. This all looks great, one addition to the docs, but other than that great job!

File: lib/plugins/aws/deploy/compile/events/apiGateway/README.md [hunk 2]
```diff
@@ -96,6 +96,25 @@ functions:
...
+      - http:
+          method: get
+          path: whatever
+          request:
+            passThrough: NEVER
+```
+
+There are currently 3 options for this: NEVER, WHEN_NO_MATCH, WHEN_NO_TEMPLATES.  See the [api gateway documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#integration-passthrough-behaviors) for detailed descriptions of these options.
+
+**Note:** API Gateway docs refer to "WHEN_NO_TEMPLATE" (singular), but this will fail during creation as the actual value should be "WHEN_NO_TEMPLATES" (plural)
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/README.md]: Wow! Great that you mentioned this one! 👍

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js [hunk 3]
```diff
@@ -147,6 +152,22 @@ module.exports = {
               ].join('');
               throw new this.serverless.classes.Error(errorMessage);
             }
+
+
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: If time permits you might want to remove this whitespace newline. 👍

**Comment by pmuens** [comment]
@anreved this is really great! Thank you for this PR! 👍 💯 
I love your 👀 for the details!

Looked through the code and added a few comments.

Would love to hear @HyperBrain feedback on this one!

**Comment by anreved** [comment]
@pmuens, @flomotlik  - not sure how this pulled in 16 commits - should I not have updated from master before making my changes?

**Comment by flomotlik** [comment]
@anreved updating from master definitely is great, but looks like there was an issue with the rebase? Could you either rebase again to clean it up or open a new PR?

**Comment by anreved** [comment]
This PR has been closed in favor of PR #2000.

**Pull request closed without merging**